_drafts/http4s-tutorial.md

```diff
@@ -0,0 +1,563 @@
+
+In detail, every `Request[F]` extends the `Media[F]` trait. This trait exposes many practical methods dealing with the body of a request, and the most interesting is the following:
+
+```scala
+final def as[A](implicit F: MonadThrow[F], decoder: EntityDecoder[F, A]): F[A] = ???
+```
+
+The `as` function decodes a request body as a type `A`. using an `EntityDecoder[F, A]`, which we must provide in the context as an implicit object.
+
+**The `EntityDecoder` (and its counterpart `EntityEncoder`) allows us to deal with the streaming nature of the data in an HTTP body**. In addition, decoders and encoders relate directly to the `Content-Type` declared as an HTTP Header in the request/response. Indeed, we need a different kind of decoder and encoder for every type of content.
+
+The http4s library ships with decoders and encoders for a limited type of contents, such as `String`, `File`, `InputStream`, and manages more complex contents using plugin libraries.
+
+In our example, the request contains a new director in JSON format. **To deal with JSON body, the most frequent choice is to use the Circe plugin**:
+
+```sbt
+"org.http4s" %% "http4s-circe" % 0.21.23,
+```
+
+The primary type provided by the Circe library to manipulate JSON information is the `io.circe.Json` type. Hence, the `http4s-circe` module defines the types `EntityDecoder[Json]` and `EntityEncoder[Json]`, which are all we need to translate a request and a response body into an instance of `Json`.
+
+However, `Json` type translate directly into JSON literals, such as `json"""{"firstName": "Zack", "lastName": "Snyder"}"""`, and we don't really want to deal with them. Instead, we usually want a case class to represent JSON information. In our example, the JSON representation of a director could translate in the following case class:
+
+```scala
+case class Director(firstName: String, lastName: String)
+```
+
+First, we decode the incoming request automatically into an instance of a `Json` object using the `EntityDecoder[Json]` type. However, we need to do a step beyond to obtain an object of type `Director`. In detail, Circe needs an instance of the type `io.circe.Decoder[Director]` to decode a `Json` object into a `Director` object.
+
+We can provide the instance of the `io.circe.Decoder[Director]` simply adding the import to `io.circe.generic.auto._`, which lets Circe automatically derive for us encoders and decoders. The derivation uses field names of case classes as key names in JSON objects and vice-versa. As the last step, we need to connect the type `Decoder[Director]` to the type `EntityDecoder[Json]`, and this is precisely the work that the module `http4s-circe` does for us through the function `jsonOf`:
+
+```scala
+implicit val directorDecoder: EntityDecoder[F, Director] = jsonOf[F, Director]
+```
+
+Summing up, the final form of the API looks like the following:
+
+```scala
+def directorRoutes[F[_]: Sync]: HttpRoutes[F] = {
+  val dsl = new Http4sDsl[F]{}
+  import dsl._
```

Comment:
You're right. However, I realized that it's not clear neither for me :(
I mean, if I try to add the imports of the single needed components, the compiler yells because a `Response[IO]` is created, instead of a `Response[F]`.

Can you help me to understand?

Reply:
It might be that the `dsl._` import works well because pinpoint importing (e.g. exact methods/objects) might be too frustrating. If the necessary extensions are too much to isolate, then we can just add a comment that this import adds some extension methods and implicit conversions for the magic below to work.